pulsar/consumer_partition.go

```diff
@@ -516,15 +577,27 @@ func (pc *partitionConsumer) grabConn() error {
 	initialPosition := toProtoInitialPosition(pc.options.subscriptionInitPos)
 	requestID := pc.client.rpcClient.NewRequestID()
 	cmdSubscribe := &pb.CommandSubscribe{
-		RequestId:       proto.Uint64(requestID),
-		Topic:           proto.String(pc.topic),
-		SubType:         subType.Enum(),
-		Subscription:    proto.String(pc.options.subscription),
-		ConsumerId:      proto.Uint64(pc.consumerID),
-		ConsumerName:    proto.String(pc.name),
-		InitialPosition: initialPosition.Enum(),
-		Schema:          nil,
+		Topic:                      proto.String(pc.topic),
+		Subscription:               proto.String(pc.options.subscription),
+		SubType:                    subType.Enum(),
+		ConsumerId:                 proto.Uint64(pc.consumerID),
+		RequestId:                  proto.Uint64(requestID),
+		ConsumerName:               proto.String(pc.name),
+		PriorityLevel:              nil,
+		Durable:                    proto.Bool(pc.options.subscriptionMode == durable),
+		Metadata:                   internal.ConvertFromStringMap(pc.options.metadata),
+		ReadCompacted:              proto.Bool(pc.options.readCompacted),
+		Schema:                     nil,
+		InitialPosition:            initialPosition.Enum(),
+		ReplicateSubscriptionState: proto.Bool(pc.options.replicateSubscriptionState),
 	}
+
+	pc.startMessageID = pc.clearReceiverQueue()
```

Comment:
Should we ensure that the receiver queue is only ever cleared from a single go routine? This was the reason it was done in the dispatcher. There could be a small random race condition here because the dispatcher go routine and the eventsLoop go routine will both be reading from the queueCh so when the nil message is sent there is a possibility the dispatcher will read it and then grabConn with never return.

Reply:
The `clearReceiverQueue()` is only called on reconnection from the consumer routine, so I don't think there should be any race condition here.

Reply:
@cckellogg You're right, there's a race condition there. It should be fixed with the latest commit. PTAL again.